packages/medusa/src/api-v2/store/carts/[id]/complete/route.ts

```diff
@@ -0,0 +1,34 @@
+import { completeCartWorkflow } from "@medusajs/core-flows"
+import { MedusaRequest, MedusaResponse } from "../../../../../types/routing"
+import { refetchOrder } from "../../../orders/helpers"
+import { StoreCompleteCartType } from "../../validators"
+
+export const POST = async (
+  req: MedusaRequest<StoreCompleteCartType>,
+  res: MedusaResponse
+) => {
+  const { idempotency_key: idempotencyKey } = req.validatedBody
+
+  // If the idempotencyKey is present:
+  //  - is workflow is running?
+  //    = throw error
+  //  - else
+  //   - re-run the workflow at the failed step
+
+  const { errors, result } = await completeCartWorkflow(req.scope).run({
+    input: { id: req.params.id },
+    throwOnError: false,
+  })
+
+  if (Array.isArray(errors) && errors[0]) {
+    throw errors[0].error
+  }
+
+  const order = await refetchOrder(
+    result.id,
+    req.scope,
+    req.remoteQueryConfig.fields
+  )
+
+  res.status(200).json({ order })
```

Comment:
thought(non-blocking): There will likely be failure modes where we have to return early with the cart instead of the order e.g. if the payment authorization falls through

Reply:
Shouldn't we throw with special codes in those cases?

Reply:
we do not have that feature currently to re run a workflow from a failed step, instead maybe we should have smaller workflow that takes care of the different blocks of action and we run them sequentially here with the existing idempotency key mechanism? what is the plan at the moment?

cc @carlos-r-l-rodrigues

Reply:
If we have time, I think we should go ahead and try getting this working with one workflow. We already have all the data in the workflow engine, we just have to point the orchestrator to run the flow to the right step. 

If not, like you said, we'll have to fall back to route level handling.